File: solutions/python/resistor-color-trio/2/resistor_color_trio.py
"""
Module for calculating the resistance value of a resistor based on color bands.
"""

RESISTOR_COLORS = (
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
)

value = RESISTOR_COLORS.index


def label(colors):
    """
    Calculate the resistance value of a resistor given its color bands.

    Parameters:
    colors (list of str): The list of color bands on the resistor.

    Returns:
    str: The resistance value as a string with appropriate prefix.
    """
    first_color, second_color, third_color = colors[:3]
    resistance_value = (
        10 * value(first_color) + value(second_color)
    ) * 10 ** value(third_color)

    prefixes = (
        (1_000_000_000, "giga"),
        (1_000_000, "mega"),
        (1_000, "kilo"),
    )
    prefix = ""

    for threshold, name in prefixes:
        if resistance_value >= threshold:
            resistance_value /= threshold
            prefix = name
            break

    return f"{int(resistance_value)} {prefix}ohms"

File: solutions/elixir/resistor-color-trio/2/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @color_values %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label([first | [second | [third | _]]]) do
    value = (@color_values[first] * 10 + @color_values[second]) * Integer.pow(10, @color_values[third])
    unit = if value < 1000, do: :ohms, else: if value < 1000000, do: :kiloohms, else: if value < 1000000000, do: :megaohms, else: :gigaohms
    case unit do
      :ohms -> {value, unit}
      :kiloohms -> {div(value, 1000), unit}
      :megaohms -> {div(value, 1000000), unit}
      :gigaohms -> {div(value, 1000000000), unit}
    end
  end
end
